### DISCUSSION TITLE
topNavigation with Clusters - no drop-down?

### DISCUSSION THREAD

**Discussion by nielspeen** [filamentphp/filament#11625] [Help]
### Package

Panel builder

### Package Version

v3.2.40

### How can we help you?

I'm using both Clusters and topNavigation. When topNavigation is enabled, Clusters nicely move to the top, just like regular NavigationItems and NavigationGroups.

But unlike the NavigationGroups, Clusters do present their items as drop-downs. Is there any way to make the clusters behave like NavigationGroups when topNavigation is enabled?

IMHO this would make for the perfect navigation. Clusters as the top, as drop-downs, with the current cluster showing it's items at the left, just like it does currently.

**Comment by mahmoud-757** [comment]
If you find out how to do it please let me know

**Reply by leandrocfe** [reply]
Honestly, I didn't understand what you want to achieve. Could you provide more information or screenshots to clarify your goals?

**Reply by mahmoud-757** [reply]
Something like this
![Artboard 1](https://github.com/filamentphp/filament/assets/147031958/ebb5b082-1fd2-4172-98a7-6c68b6aa0319)

**Reply by leandrocfe** [reply]
You can use this in your resource

```php
protected static SubNavigationPosition $subNavigationPosition = SubNavigationPosition::Top;
```
to change the resource navigation position

![image](https://github.com/filamentphp/filament/assets/3833889/f04d6c36-beb9-4ff9-a217-7284c985a1be)

**Reply by mahmoud-757** [reply]
Yes but not in the create/edit/view page it takes the full page width an removes the sub nav entirely, or am I missing something?

**Reply by leandrocfe** [reply]
Correct, it is available in the ListPage. If you want to add in the create/edit, you should add this method on each page

```php
public function getSubNavigation(): array
{
    if (filled($cluster = static::getCluster())) {
        return $this->generateNavigationItems($cluster::getClusteredComponents());
    }

    return [];
}
```

**Reply by mahmoud-757** [reply]
Thanks that is exactly what I want, it would be great if there was a global way to do it since there I have a lot of resource pages, but not complaining xD

**Reply by leandrocfe** [reply]
You can create a trait and use it in all pages.
Hey @nielspeen, check if it works in your case please.

**Reply by nielspeen** [reply]
@leandrocfe it does what @mahmoud-757 requested, but not what was in my original request. I'm looking for the cluster to generate a drop-down menu when using topNavigation.

If you enable topNavigation for your panel, and use both Clusters and traditional NavigationGroups you'll see what I mean. The NavigationGroups appear as drop-down menus, but the Clusters do not. (Or you could say that they do, but their items are missing).

**Reply by leandrocfe** [reply]
Share a screenshot what you need please

**Reply by nielspeen** [reply]
<img width="1244" alt="Screenshot 2024-03-05 at 08 04 16" src="https://github.com/filamentphp/filament/assets/5564176/0541535b-a5fc-45c9-ab98-3c1e4d387121">

At the top, on the very right, you see a NavigationGroup **Other**, with all of its NavigationItems. It appears as a drop-down.

All the other options at the top (Clients, .., System) are Clusters. For these Clusters, the items are not shown as a drop-down.

For the currently opened Cluster **System**, displaying all the items as a drop-down may be redundant as they're already on the left of the screen. But for all other Clusters it would be very helpful to have the items appear as drop-downs.

**Reply by leandrocfe** [reply]
They are different. I believe you cannot display the cluster subnavigation as a dropdown, unless you are utilizing `SubNavigationPosition::Top`

![image](https://github.com/filamentphp/filament/assets/3833889/c137e5b6-2d04-4397-a07f-fcdbd9489e35)

**Reply by nielspeen** [reply]
Yeah, that's unfortunate.

Looks like Resource.php can actually do exactly what I want, but it's disabled for Clusters:

```php
public static function registerNavigationItems(): void
    {
        if (filled(static::getCluster())) {
            return;
        }

        if (! static::shouldRegisterNavigation()) {
            return;
        }

        if (! static::canAccess()) {
            return;
        }

        Filament::getCurrentPanel()
            ->navigationItems(static::getNavigationItems());
    }
```

Changing that first _if_ to:

```php
if (filled(static::getCluster()) && ! Filament::getCurrentPanel()->hasTopNavigation()) {
    return;
}
```

I get exactly what I want.
